src/OpenDebugAD7/TracepointManager.cs

```diff
@@ -0,0 +1,229 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.VisualStudio.Debugger.Interop;
+using Microsoft.VisualStudio.Debugger.Interop.DAP;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text.RegularExpressions;
+
+namespace OpenDebugAD7
+{
+    internal class TracepointManager
+    {
+        private readonly object _lock = new object();
+
+        private readonly Regex regex;
+        private readonly Dictionary<uint, string> m_breakpointLogMessages;
+        private readonly enum_DEBUGPROP_INFO_FLAGS propertyInfoFlags;
+        private readonly enum_EVALFLAGS flags;
+
+        internal TracepointManager()
+        {
+            m_breakpointLogMessages = new Dictionary<uint, string>();
+
+            // Matches strings that are in { } 
+            // or upper case strings that begin with $.
+            regex = new Regex("\\$[A-Z]+|{.*?}", RegexOptions.Compiled);
+
+            propertyInfoFlags = enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_NAME |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_VALUE |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_TYPE |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_ATTRIB |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_PROP |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_FULLNAME |
+                (enum_DEBUGPROP_INFO_FLAGS)enum_DEBUGPROP_INFO_FLAGS110.DEBUGPROP110_INFO_FORCE_REAL_FUNCEVAL |
+                (enum_DEBUGPROP_INFO_FLAGS)enum_DEBUGPROP_INFO_FLAGS110.DEBUGPROP110_INFO_NOSIDEEFFECTS;
+
+            flags = enum_EVALFLAGS.EVAL_RETURNVALUE |
+                enum_EVALFLAGS.EVAL_NOEVENTS |
+                (enum_EVALFLAGS)enum_EVALFLAGS110.EVAL110_FORCE_REAL_FUNCEVAL |
+                enum_EVALFLAGS.EVAL_NOSIDEEFFECTS;
+
+        }
+
+        internal void Add(uint breakpointId, string logMessage)
+        {
+            lock (_lock)
+            {
+                m_breakpointLogMessages.Add(breakpointId, logMessage);
```

Comment:
>logMessage [](start = 58, length = 10)

Do you want to parse and validate the log message up front?

Reply:
I do this now in CreateTracepoint(). Do if it fails do we just not bind the tracepoint?